src/kernel/binary_reduce.cc

```diff
@@ -0,0 +1,506 @@
+          reducer, op, graph,
+          lhs, rhs,
+          lhs_data, rhs_data, out_data,
+          lhs_mapping, rhs_mapping, out_mapping);
+    }
+  }
+}
+
+DGL_REGISTER_GLOBAL("kernel._CAPI_DGLKernelBinaryOpReduce")
+.set_body([] (DGLArgs args, DGLRetValue* rv) {
+    std::string reducer = args[0];
+    std::string op = args[1];
+    GraphHandle ghdl = args[2];
+    int lhs = args[3];
+    int rhs = args[4];
+    NDArray lhs_data = args[5];
+    NDArray rhs_data = args[6];
+    NDArray out_data = args[7];
+    NDArray lhs_mapping = args[8];
+    NDArray rhs_mapping = args[9];
+    NDArray out_mapping = args[10];
+
+    GraphInterface* gptr = static_cast<GraphInterface*>(ghdl);
+    const ImmutableGraph* igptr = dynamic_cast<ImmutableGraph*>(gptr);
+    CHECK(igptr) << "Invalid graph object argument. Must be an immutable graph.";
+    BinaryOpReduce(reducer, op, igptr,
+        static_cast<binary_op::Target>(lhs), static_cast<binary_op::Target>(rhs),
+        lhs_data, rhs_data, out_data,
+        lhs_mapping, rhs_mapping, out_mapping);
+  });
+
+void BackwardLhsBinaryOpReduce(
+    const std::string& reducer,
+    const std::string& op,
+    const ImmutableGraph* graph,
+    binary_op::Target lhs, binary_op::Target rhs,
+    NDArray lhs_mapping,
+    NDArray rhs_mapping,
+    NDArray out_mapping,
+    NDArray lhs_data,
+    NDArray rhs_data,
+    NDArray out_data,
+    NDArray grad_out_data,
+    NDArray grad_lhs_data) {
+  const auto& ctx = graph->Context();
+  // sanity check
+  CheckCtx(ctx,
+      {lhs_data, rhs_data, out_data, grad_out_data, grad_lhs_data,
+       lhs_mapping, rhs_mapping, out_mapping},
+      {"lhs_data", "rhs_data", "out_data", "grad_out_data", "grad_lhs_data",
+       "lhs_mapping", "rhs_mapping", "out_mapping"});
+  CheckIdArray(graph->NumBits(),
+      {lhs_mapping, rhs_mapping, out_mapping},
+      {"lhs_mapping", "rhs_mapping", "out_mapping"});
+  // Switch order for commutative operation
+  if (NeedSwitchOrder(op, lhs, rhs)) {
+    BackwardRhsBinaryOpReduce(reducer, op, graph,
+        rhs, lhs,
+        rhs_mapping, lhs_mapping, out_mapping,
+        rhs_data, lhs_data, out_data,
+        grad_out_data, grad_lhs_data);
+  } else {
+    if (HasBcast(lhs_data, rhs_data)) {
+      BcastInfo info = CalcBcastInfo(lhs_data, rhs_data);
+      DGL_XPU_SWITCH(ctx.device_type, BackwardBinaryReduceBcastImpl,
+          info, reducer, op, graph,
+          lhs, rhs,
+          lhs_mapping, rhs_mapping, out_mapping,
+          lhs_data, rhs_data, out_data, grad_out_data,
+          grad_lhs_data, utils::NoneArray());
+    } else {
+      DGL_XPU_SWITCH(ctx.device_type, BackwardBinaryReduceImpl,
+          reducer, op, graph,
+          lhs, rhs,
+          lhs_mapping, rhs_mapping, out_mapping,
+          lhs_data, rhs_data, out_data, grad_out_data,
+          grad_lhs_data, utils::NoneArray());
+    }
+  }
+}
+
+DGL_REGISTER_GLOBAL("kernel._CAPI_DGLKernelBackwardLhsBinaryOpReduce")
```

Comment:
why do you need to break it into two CAPIs?

Reply:
Because it is possible that only one gradient is needed. For example, in `u_mul_e` case, if e is a fixed edge feature that does not require gradient, then only BackwardLhsBinaryOpReduce will be called.